projects/driver_controls/src/ads1015.c

```diff
@@ -0,0 +1,169 @@
+  if (channel >= NUM_ADS1015_CHANNELS) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  return STATUS_CODE_OK;
+}
+
+// Initiates a read as outlined in section 8.5.3 of the datasheet
+static StatusCode prv_read(I2CPort i2c_port, uint8_t reg, uint8_t *data) {
+  status_ok_or_return(i2c_write(i2c_port, ADS1015_I2C_ADDRESS, &reg, 1));
+  status_ok_or_return(i2c_read(i2c_port, ADS1015_I2C_ADDRESS, data, 2));
+
+  return STATUS_CODE_OK;
+}
+
+// Initiates a write as outlined in section 8.5.3 of the datasheet
+static StatusCode prv_write(I2CPort i2c_port, uint8_t reg, uint8_t *data) {
+  // For writes, the input data must be sent in the same frame as the address pointer register
+  uint8_t write_data[] = { reg, data[0], data[1] };
+  status_ok_or_return(i2c_write(i2c_port, ADS1015_I2C_ADDRESS, write_data, 3));
+
+  return STATUS_CODE_OK;
+}
+
+static void prv_interrupt_handler(GPIOAddress *address, void *context) {
+  ADS1015Register reg;
+
+  // Obtain ADC readings
+  status_ok_or_return(prv_read(s_i2c_port, ADS1015_CONVERSION_REGISTER, reg.data));
+  s_interrupts[s_current_channel].reading = (reg.data[0] << 4) | (reg.data[1] >> 4);
+
+  if (s_interrupts[s_current_channel].callback != NULL) {
+    s_interrupts[s_current_channel].callback(s_current_channel,
+                                              s_interrupts[s_current_channel].context);
+  }
+
+  // Update so that the ADC reads from the next channel
+  s_current_channel = (s_current_channel + 1) % NUM_ADS1015_CHANNELS;
+
+  // Obtain the current value for the config register
+  status_ok_or_return(prv_read(s_i2c_port, ADS1015_CONFIG_REGISTER, reg.data));
+
+  // Set bits 12-14 based on the current channel
+  switch (s_current_channel) {
+    case ADS1015_CHANNEL_0:
+      reg.data[0] = ADS1015_CONFIG_MUX_AIN0(reg.data[0]);
+      break;
+    case ADS1015_CHANNEL_1:
+      reg.data[0] = ADS1015_CONFIG_MUX_AIN1(reg.data[0]);
+      break;
+    case ADS1015_CHANNEL_2:
+      reg.data[0] = ADS1015_CONFIG_MUX_AIN2(reg.data[0]);
+      break;
+    case ADS1015_CHANNEL_3:
+      reg.data[0] = ADS1015_CONFIG_MUX_AIN3(reg.data[0]);
+      break;
+  }
+
+  status_ok_or_return(prv_write(s_i2c_port, ADS1015_CONFIG_REGISTER, reg.data));
+
+  return STATUS_CODE_OK;
+}
+
+StatusCode ads1015_init(I2CPort i2c_port, GPIOAddress address) {
+  s_i2c_port = i2c_port;
+
+  // Configure the given GPIO address as a conversion ready pin
+  GPIOSettings gpio_settings = { .direction = GPIO_DIR_IN, .alt_function = GPIO_ALTFN_NONE };
+  InterruptSettings it_settings = { INTERRUPT_TYPE_INTERRUPT, INTERRUPT_PRIORITY_NORMAL };
+
+  gpio_init_pin(&address, &gpio_settings);
+  gpio_it_register_interrupt(&address, &it_settings, INTERRUPT_EDGE_FALLING,
+                              prv_interrupt_handler, NULL);
+
+  // Reset the internal registers to their default values
+  uint8_t reset = ADS1015_RESET_BYTE;
+  status_ok_or_return(i2c_write(s_i2c_port, 0, &reset, 1));
```

Comment:
If you're sending reset packets to the general call address, you'll be resetting other slaves on the bus as well.